# Sparse Project VCF (spVCF)

Project VCF (pVCF; aka multi-sample VCF) is the prevailing file format for small genetic variants discovered by cohort sequencing. It encodes a two-dimensional matrix with variant sites down the rows and study participants across the columns, filled in with all the genotypes and associated QC measures (read depths, genotype likelihoods, etc.). Large cohorts harbor many rare variants, implying a sparse genotype matrix composed largely of reference-homozygous or non-called cells. But the dense pVCF format encodes this inefficiently, growing super-linearly with the cohort size.

To quantify this problem, in recent experiments with up to *N*=50K human exomes, 96% of pVCF sites had non-reference allele frequency below 0.1%, while doubling *N* from 25K also increased the number of sites by 43% (cf. Table 2 of the [GLnexus preprint](https://www.biorxiv.org/content/early/2018/06/11/343970)).

spVCF explores a minimal evolution of VCF to encode the matrix sparsely. This aims to measure what can be gained with a modest change; much less sophisticated than other efforts to address VCF's density and other shortcomings, but perhaps more palatable to existing VCF consumers.

Starting from text pVCF, spVCF:

1. Replaces the contents of a cell with `"` if they're identical to the cell *above*. Even with several numeric QC fields, such repetition is common in large pVCF files generated by merging many genome VCF (gVCF) files, or similar intermediates, which convey reference coverage in lengthy bands.
2. Run-length encodes these across the rows, so for example a *horizontal* run of 42 quotes is written `"42`
3. *Optional*: suppresses fields other than genotype `GT` and read depth `DP` from cells indicating no non-reference reads `AD=*,0`. Also rounds `DP` down to a power of two in these cells (0, 1, 2, 4, 8, 16, ...), increasing compressibility while preserving information to distinguish reference-homozygous from unknown genotypes.

[This slide deck](https://docs.google.com/presentation/d/1twKjLygz8tmq1wb3dqKmXFArUvK4zDoiwvfpAdVbRJo/edit?usp=sharing) has further detail and examples, and preliminary test results are summarized in [doc/compression_results.md](https://github.com/mlin/spVCF/blob/master/doc/compression_results.md). Briefly, with *N*=50K exomes, spVCF with the optional entropy suppression delivers >15X size reduction for bgzip-compressed pVCF, scaling much more gently with *N*.

## Build and test

[![Build Status](https://travis-ci.org/mlin/spVCF.svg?branch=master)](https://travis-ci.org/mlin/spVCF)

```
cmake . && make
ctest -V
```

## Usage

`spvcf encode` and `spvcf decode` encode existing pVCF to spVCF and vice versa. The input and output streams are uncompressed VCF text, so you may wish to arrange a pipe with [de]compression programs like `bgzip`.

```
spvcf encode [options] [in.vcf|-]
Reads VCF text from standard input if filename is empty or -

Options:
  -o,--output out.spvcf  Write to out.spvcf instead of standard output
  -p,--period P          Ensure checkpoints (full dense rows) at this period or less (default: 1000)
  -q,--quiet             Suppress statistics printed to standard error
  -h,--help              Show this help message

Lossy transformation to increase compression: 
  -S,--squeeze           Truncate cells to GT:DP, with DP rounded down to a power of two, if: 
                         - AD is present and indicates zero read depth for alternate alleles; OR
                         - VR is present and zero
                         Reorders fields within all cells.
```

```
spvcf decode [options] [in.spvcf|-]
Reads spVCF text from standard input if filename is empty or -

Options:
  -o,--output out.vcf  Write to out.vcf instead of standard output
  -q,--quiet           Suppress statistics printed to standard error
  -h,--help            Show this help message
```

Examples:

```
$ ./spvcf encode my.vcf > my.spvcf
$ bgzip -dc my.vcf.gz | ./spvcf encode | bgzip -c > my.spvcf.gz
$ bgzip -dc my.spvcf.gz | ./spvcf decode > my.decoded.vcf
```

### Tabix slicing

If the regular `bgzip` and `tabix` utilities are used to block-compress and index a spVCF file, then `spvcf tabix` can take a genomic range slice from it, extracting spVCF which decodes standalone. (The regular `tabix` utility generates the index, but using it to take the slice would yield a broken fragment.) Internally, this entails decoding a small bit of the spVCF, determined by the `spvcf encode --period` option.

```
spvcf tabix [options] in.spvcf.gz chr1:1000-2000 [chr2 ...]
Requires tabix index present e.g. in.spvcf.gz.tbi. Includes all header lines.

Options:
  -o,--output out.spvcf  Write to out.spvcf instead of standard output
  -h,--help              Show this help message
```

Example:

```
$ ./spvcf encode my.vcf | bgzip -c > my.spvcf.gz
$ tabix my.spvcf.gz
$ ./spvcf tabix my.spvcf.gz chr21:5143000-5219900 | ./spvcf decode > slice.vcf
```
