# Sparse Project VCF (spVCF)

Project VCF (pVCF; aka multi-sample VCF) is the prevailing file format for small genetic variants discovered by cohort sequencing. It encodes a two-dimensional matrix with variant sites down the rows and study participants across the columns, filled in with all the genotypes and associated QC measures (read depths, genotype likelihoods, etc.). Large cohorts harbor many rare variants, implying a sparse genotype matrix composed largely of reference-homozygous or non-called cells. But the dense pVCF format encodes this inefficiently, growing super-linearly with the cohort size.

spVCF explores a minimal evolution of VCF to encode the matrix sparsely. This aims to measure what can be gained with a modest change; much less sophisticated than other efforts to address VCF's density and other shortcomings, but perhaps more palatable to existing VCF consumers.

Starting from text pVCF, spVCF:

1. Replaces the contents of a cell with `"` if they're identical to the cell *above*. Even with several numeric QC fields, such repetition is common in large pVCF files generated by merging many genome VCF (gVCF) files, or similar intermediates, which convey reference coverage in lengthy bands.
2. Run-length encodes these across the rows, so for example a *horizontal* run of 42 quotes is written `"42`
3. *Optionally*, discards the QC measures in cells with only reference or non-called genotypes (prior to steps #1 and #2, amplifying their effectiveness)

That's all! Drawbacks: breaking change to VCF, and complicates random access into the matrix.

## Build and test

[![Build Status](https://travis-ci.org/mlin/spVCF.svg?branch=master)](https://travis-ci.org/mlin/spVCF)

```
cmake -DCMAKE_BUILD_TYPE=Release . && make
ctest -V
```

## Usage

The `spvcf` executable encodes an existing pVCF to spVCF, or conversely decodes spVCF to pVCF. The input and output streams are uncompressed VCF text, so you may wish to arrange a pipe with [de]compression programs like `bgzip`.

```
$ ./spvcf -h
spvcf [options] [in.vcf|-]
Read from standard input if input filename is empty or -
Options:
  -o,--output out.spvcf  Write to out.spvcf instead of standard output
  -d,--decode            Decode from the sparse format instead of encoding to
  -S,--squeeze           Discard QC measures from cells with no ALT allele called
  -p,--checkpoint-period Ensure checkpoints (full dense row) at this period or less [1000]
  -q,--quiet             Suppress statistics printed to standard error
  -h,--help              Show this usage message
```

Examples:

```
$ ./spvcf my.vcf > my.spvcf
$ bgzip -dc my.vcf.gz | ./spvcf | bgzip -c > my.spvcf.gz
$ bgzip -dc my.spvcf.gz | ./spvcf -d > my.decoded.vcf
```
