# Sparse Project VCF (spVCF)

Project VCF (pVCF; aka multi-sample VCF) is the prevailing file format for small genetic variants discovered by cohort sequencing. It encodes a two-dimensional matrix with variant sites down the rows and study participants across the columns, filled in with all the genotypes and associated QC measures (read depths, genotype likelihoods, etc.). Large cohorts harbor many rare variants, implying a sparse genotype matrix composed largely of reference-homozygous or non-called cells. But the dense pVCF format encodes this inefficiently, growing super-linearly with the cohort size.

To quantify this problem, in recent experiments with up to *N*=50K human exomes, 96% of pVCF sites have non-reference allele frequency below 0.1%, while doubling *N* from 25K also increased the number of sites by 43% (cf. Table 2 of the [GLnexus preprint](https://www.biorxiv.org/content/early/2018/06/11/343970)).

spVCF explores a minimal evolution of VCF to encode the matrix sparsely. This aims to measure what can be gained with a modest change; much less sophisticated than other efforts to address VCF's density and other shortcomings, but perhaps more palatable to existing VCF consumers.

Starting from text pVCF, spVCF:

1. Replaces the contents of a cell with `"` if they're identical to the cell *above*. Even with several numeric QC fields, such repetition is common in large pVCF files generated by merging many genome VCF (gVCF) files, or similar intermediates, which convey reference coverage in lengthy bands.
2. Run-length encodes these across the rows, so for example a *horizontal* run of 42 quotes is written `"42`
3. *Optional*: suppresses fields other than genotype `GT` and read depth `DP` from cells indicating no non-reference reads `AD=*,0`. Also rounds `DP` down to a power of two in these cells (0, 1, 2, 4, 8, 16, ...), increasing compressibility while preserving information to distinguish reference-homozygous from unknown genotypes.

[This slide deck](https://docs.google.com/presentation/d/1twKjLygz8tmq1wb3dqKmXFArUvK4zDoiwvfpAdVbRJo/edit?usp=sharing) has further detail and examples, and preliminary test results are summarized in [doc/compression_results.md](https://github.com/mlin/spVCF/blob/master/doc/compression_results.md). Briefly, with *N*=50K exomes, spVCF with the optional entropy suppression delivers >15X size reduction for bgzip-compressed pVCF, scaling much more gently with *N*.

## Build and test

[![Build Status](https://travis-ci.org/mlin/spVCF.svg?branch=master)](https://travis-ci.org/mlin/spVCF)

```
cmake -DCMAKE_BUILD_TYPE=Release . && make
ctest -V
```

## Usage

The `spvcf` executable encodes an existing pVCF to spVCF, or conversely decodes spVCF to pVCF. The input and output streams are uncompressed VCF text, so you may wish to arrange a pipe with [de]compression programs like `bgzip`.

```
spvcf [options] [in.vcf|-]
Reads VCF text from standard input if filename is empty or -

Options:
  -o,--output out.spvcf  Write to out.spvcf instead of standard output
  -d,--decode            Decode from the sparse format instead of encoding to
  -p,--period P          Ensure checkpoints (full dense rows) at this period or less (default: 1000)
  -q,--quiet             Suppress statistics printed to standard error
  -h,--help              Show this usage message

Lossy transformation to increase compression:
  -S,--squeeze           Truncate cells to GT:DP, with DP rounded down to a power of two, if:
                         - AD is present and indicates zero read depth for alternate alleles; OR
                         - VR is present and zero
                         Reorders fields within all cells.
```

Examples:

```
$ ./spvcf my.vcf > my.spvcf
$ bgzip -dc my.vcf.gz | ./spvcf | bgzip -c > my.spvcf.gz
$ bgzip -dc my.spvcf.gz | ./spvcf -d > my.decoded.vcf
```
